src/com/reco1l/legacy/AccessibilityDetector.kt

```diff
@@ -0,0 +1,93 @@
+package com.reco1l.legacy
+
+import android.accessibilityservice.AccessibilityServiceInfo
+import android.accessibilityservice.AccessibilityServiceInfo.CAPABILITY_CAN_PERFORM_GESTURES
+import android.accessibilityservice.AccessibilityServiceInfo.FEEDBACK_ALL_MASK
+import android.app.AlertDialog
+import android.content.Context.ACCESSIBILITY_SERVICE
+import android.content.Intent
+import android.provider.Settings
+import android.view.accessibility.AccessibilityManager
+import ru.nsu.ccfit.zuev.osu.MainActivity
+import ru.nsu.ccfit.zuev.osuplus.R
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+
+object AccessibilityDetector
+{
+
+
+    /**
+     * `true` if an accessibility services with the ability of perform gestures is detected.
+     */
+    @JvmStatic
+    var isIllegalServiceDetected = true
+        private set
+
+    /**
+     * The alert that notifies the user which services are not allowed
+     */
+    private var alert: AlertDialog? = null
+
+
+    @JvmStatic
+    fun start(context: MainActivity) = Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate({
+
+        // Getting the accessibility manager.
+        val manager = context.getSystemService(ACCESSIBILITY_SERVICE) as AccessibilityManager
+
+        // Filtering services that can perform gestures.
+        val illegalServices = manager.getEnabledAccessibilityServiceList(FEEDBACK_ALL_MASK).filter {
+
+            CAPABILITY_CAN_PERFORM_GESTURES and it.capabilities == CAPABILITY_CAN_PERFORM_GESTURES
+        }
+
+        isIllegalServiceDetected = illegalServices.isNotEmpty()
+
+        context.runOnUiThread {
+
+            if (isIllegalServiceDetected)
+            {
+                if (alert == null)
+                    alert = showAlert(context, illegalServices)
+
+                return@runOnUiThread
+            }
+
+            alert?.dismiss()
+            alert = null
+        }
+
+    }, 0, 1000, TimeUnit.MILLISECONDS)!!
+
+
+    private fun showAlert(context: MainActivity, services: List<AccessibilityServiceInfo>) = AlertDialog.Builder(context).apply {
+
+        val message = buildString {
+
+            append(context.getString(R.string.accessibility_detector_message))
+            appendLine()
+
+            for (service in services) service.resolveInfo.serviceInfo.let {
+                appendLine()
+                append(it.packageName)
+            }
+        }
+
+        setTitle(R.string.accessibility_detector_title)
+        setMessage(message)
+        setCancelable(false)
+        setNeutralButton(R.string.accessibility_detector_settings) { _, _ ->
+
+            val intent = Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS)
```

Comment:
Does the existence of this action not need to be checked? It goes against the [Android docs](https://developer.android.com/reference/android/provider/Settings#ACTION_ACCESSIBILITY_SETTINGS).

Reply:
If there's "no match activity" like the docs says, it'll just do nothing.

Reply:
Hmm in that case we can try-catch it, and show a toast in case it fails.

Also a proper solution should be check if that activity exists and add the button according to that but I don't know if it's possible.

Reply:
From [this](https://developer.android.com/guide/components/intents-common.html) documentation:

> If there are no apps on the device that can receive an implicit intent, an app crashes when it calls [startActivity()](https://developer.android.com/reference/android/content/Context#startActivity(android.content.Intent)). To first verify that an app exists to receive the intent, call [resolveActivity()](https://developer.android.com/reference/android/content/Intent#resolveActivity(android.content.pm.PackageManager)) on your [Intent](https://developer.android.com/reference/android/content/Intent) object. If the result is non-null, there is at least one app that can handle the intent, and it's safe to call startActivity(). If the result is null, don't use the intent and, if possible, disable the feature that invokes the intent.

Reply:
I will do it, should not be hard.